File: Frontend/src/services/chatService.js
import { api } from "./api"

function sessionToFrontend(s) {
  return {
    id: s.id,
    title: s.title || "New Chat",
    personalityType: s.personality_type || "empathetic",
    createdAt: s.created_at,
    updatedAt: s.updated_at,
    lastMessageAt: s.last_message_at,
  }
}

function messageToFrontend(m) {
  return {
    id: m.id,
    sessionId: m.session_id,
    role: m.role,
    content: m.content,
    journalContext: m.journal_context || null,
    personalityMode: m.personality_mode || null,
    createdAt: m.created_at,
  }
}

export const chatService = {
  async getSessions() {
    const data = await api.get("/chats")
    return { sessions: data.sessions.map(sessionToFrontend) }
  },

  async getSession(id) {
    const data = await api.get(`/chats/${id}`)
    return sessionToFrontend(data.session)
  },

  async renameSession(id, title) {
    const data = await api.put(`/chats/${id}`, { title })
    return sessionToFrontend(data.session)
  },

  async deleteSession(id) {
    return api.delete(`/chats/${id}`)
  },

  async getMessages(sessionId) {
    const data = await api.get(`/chats/${sessionId}/messages`)
    return { messages: data.messages.map(messageToFrontend) }
  },
}

File: Backend/app/services/chat_service.py
from app.extensions import db
from app.models.chat import ChatSession, ChatMessage


class ChatService:
    @staticmethod
    def get_sessions(user_id):
        return (
            ChatSession.query
            .filter_by(user_id=user_id)
            .order_by(ChatSession.last_message_at.desc().nullslast())
            .all()
        )

    @staticmethod
    def get_session(session_id, user_id):
        return ChatSession.query.filter_by(id=session_id, user_id=user_id).first()

    @staticmethod
    def create_session(user_id, title):
        session = ChatSession(user_id=user_id, title=title.strip())
        db.session.add(session)
        db.session.commit()
        return session

    @staticmethod
    def rename_session(session_id, user_id, title):
        session = ChatSession.query.filter_by(id=session_id, user_id=user_id).first()
        if not session:
            return None
        session.title = title.strip()
        db.session.commit()
        return session

    @staticmethod
    def delete_session(session_id, user_id):
        session = ChatSession.query.filter_by(id=session_id, user_id=user_id).first()
        if not session:
            return False
        db.session.delete(session)
        db.session.commit()
        return True

    @staticmethod
    def get_messages(session_id, user_id):
        session = ChatSession.query.filter_by(id=session_id, user_id=user_id).first()
        if not session:
            return None
        return (
            ChatMessage.query
            .filter_by(session_id=session_id)
            .order_by(ChatMessage.created_at)
            .all()
        )

    @staticmethod
    def create_message(session_id, user_id, role, content, personality_mode=None):
        session = ChatSession.query.filter_by(id=session_id, user_id=user_id).first()
        if not session:
            return None
        message = ChatMessage(
            session_id=session_id,
            role=role,
            content=content.strip(),
            personality_mode=personality_mode,
        )
        db.session.add(message)
        session.updated_at = db.func.now()
        session.last_message_at = db.func.now()
        db.session.commit()
        return message

    @staticmethod
    def set_personality(session_id, user_id, personality):
        """Update the personality type for a chat session."""
        session = ChatSession.query.filter_by(id=session_id, user_id=user_id).first()
        if not session:
            return None
        old = session.personality_type
        session.personality_type = personality
        db.session.commit()
        return session
